go.mod

```diff
@@ -50,6 +50,7 @@ require (
 	k8s.io/gengo v0.0.0-20190308184658-b90029ef6cd8 // indirect
 	k8s.io/klog v0.2.0
 	k8s.io/kube-openapi v0.0.0-20190306001800-15615b16d372 // indirect
+	k8s.io/kubernetes v1.13.3
```

Comment:
Importing `k8s.io/kubernetes` should be avoided if possible, why do we need this?

Reply:
Oh for kubeadm I see

Reply:
This most likely comes from importing `kubeadm` types, I'm not up to speed with go modules, is there a way to scope the import only to the sub package we need here without importing the whole shebang? (or maybe that's what happens already)

Reply:
It's `kubeadm` + all it's transitive deps which is not the whole thing but still significant. Doesn't look like there's a reasonable work around for now though. Staging `kubadm` would address this, that discussion is on-going.

Reply:
Go modules strips unused code paths. If you look in the actual vendor you'll see this:

```shell
$ find vendor/k8s.io/kubernetes/ -type f
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/register.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/types.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/bootstraptokenstring.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/doc.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/register.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults_unix.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults_windows.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/zz_generated.defaults.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/bootstraptokenstring.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/BUILD
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/zz_generated.conversion.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/v1beta1/zz_generated.deepcopy.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/BUILD
vendor/k8s.io/kubernetes//cmd/kubeadm/app/apis/kubeadm/zz_generated.deepcopy.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/constants/constants.go
vendor/k8s.io/kubernetes//cmd/kubeadm/app/constants/BUILD
vendor/k8s.io/kubernetes//LICENSE
vendor/k8s.io/kubernetes//pkg/apis/core/register.go
vendor/k8s.io/kubernetes//pkg/apis/core/taint.go
vendor/k8s.io/kubernetes//pkg/apis/core/types.go
vendor/k8s.io/kubernetes//pkg/apis/core/annotation_key_constants.go
vendor/k8s.io/kubernetes//pkg/apis/core/field_constants.go
vendor/k8s.io/kubernetes//pkg/apis/core/resource.go
vendor/k8s.io/kubernetes//pkg/apis/core/objectreference.go
vendor/k8s.io/kubernetes//pkg/apis/core/doc.go
vendor/k8s.io/kubernetes//pkg/apis/core/OWNERS
vendor/k8s.io/kubernetes//pkg/apis/core/BUILD
vendor/k8s.io/kubernetes//pkg/apis/core/zz_generated.deepcopy.go
vendor/k8s.io/kubernetes//pkg/apis/core/toleration.go
vendor/k8s.io/kubernetes//pkg/apis/core/json.go
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/register.go
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/types.go
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/doc.go
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/OWNERS
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/BUILD
vendor/k8s.io/kubernetes//pkg/proxy/apis/config/zz_generated.deepcopy.go
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/register.go
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/types.go
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/doc.go
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/OWNERS
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/BUILD
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/helpers.go
vendor/k8s.io/kubernetes//pkg/kubelet/apis/config/zz_generated.deepcopy.go
vendor/k8s.io/kubernetes//pkg/registry/core/service/allocator/interfaces.go
vendor/k8s.io/kubernetes//pkg/registry/core/service/allocator/utils.go
vendor/k8s.io/kubernetes//pkg/registry/core/service/allocator/BUILD
vendor/k8s.io/kubernetes//pkg/registry/core/service/allocator/bitmap.go
vendor/k8s.io/kubernetes//pkg/registry/core/service/ipallocator/BUILD
vendor/k8s.io/kubernetes//pkg/registry/core/service/ipallocator/allocator.go
```

Reply:
Hi @andrewsykim and @frapposelli,

In fact, `go mod` strips things **too** well. That's why **this** exists in the [`Makefile`](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/Makefile#L88-L101):

```make
# Regenerating vendor cannot happen in a symlink due to the way certain Go
# commands traverse the file structure. Fore more information please see
# https://github.com/golang/go/issues/17451.
vendor: export SHELL_WITH_GOPATH=0
vendor: export GO111MODULE=on
vendor:
	go mod tidy -v
	go mod vendor -v
	go mod verify
	_src="$$(go list -f '{{.Dir}}' sigs.k8s.io/cluster-api/cmd/clusterctl 2>/dev/null)/../../config" && \
	_dst=./vendor/sigs.k8s.io/cluster-api/ && \
	mkdir -p "$${_dst}" && \
	cp -rf --no-preserve=mode "$${_src}" "$${_dst}"
.PHONY: vendor
```

Reply:
thanks @akutz , I'm good with canceling the hold any time

Reply:
Hi @frapposelli,

Thanks! I do need to pose a question to @figo. While the e2e test passed, I'm concerned that the static management host we use for e2e won't be using the new CRDs. I need to ask him how we account for that.